servers/rendering/rendering_device_driver.h

```diff
@@ -128,7 +128,8 @@ class RenderingDeviceDriver : public RenderingDeviceCommons {
 
 #define DEFINE_ID(m_name)                                                                             \
 	struct m_name##ID : public ID {                                                                   \
-		_ALWAYS_INLINE_ operator bool() const { return id != 0; }                                     \
+		_ALWAYS_INLINE_ explicit operator bool() const { return id != 0; }                            \
+		_ALWAYS_INLINE_ operator uint64_t() const { return id; }                                      \
```

Comment:
Why these changes?

Reply:
See the main message, it's about issues with implicit bool casts, for some context see [here](https://en.wikibooks.org/wiki/More_C++_Idioms/Safe_bool)

Reply:
Oh, good to know.

I'm now only concerned about the `operator uint64_t()`. This code was written with explicitness in mind and so all the rendering driver code assumes it has to do `.id` to get the value. Wouldn't this change make that pointless?

Reply:
> This code was written with explicitness in mind and so all the rendering driver code assumes it has to do `.id` to get the value

If we add `explicit ` to `operator bool`, then some places will compile error and so I added `operator uint64_t()` to ID.

Reply:
If some piece of code was leveraging implicit conversion to integer, that was unintended. The way this was designed, `.id` should be used explicitly. Are there many cases of that? If not a lot up to the point of having to reconsider it all, I'd rather have those unintended uses patched.

Reply:
Here's the diff needed to make it compile (at least Linux editor dev build) without that operator:
```diff
diff --git a/servers/rendering/rendering_device.cpp b/servers/rendering/rendering_device.cpp
index 35f99e187d..e0c4ad1739 100644
--- a/servers/rendering/rendering_device.cpp
+++ b/servers/rendering/rendering_device.cpp
@@ -5225,10 +5225,10 @@ uint64_t RenderingDevice::get_driver_resource(DriverResource p_resource, RID p_r
 		case DRIVER_RESOURCE_TOPMOST_OBJECT:
 			break;
 		case DRIVER_RESOURCE_COMMAND_QUEUE:
-			driver_id = main_queue;
+			driver_id = main_queue.id;
 			break;
 		case DRIVER_RESOURCE_QUEUE_FAMILY:
-			driver_id = main_queue_family;
+			driver_id = main_queue_family.id;
 			break;
 		case DRIVER_RESOURCE_TEXTURE:
 		case DRIVER_RESOURCE_TEXTURE_VIEW:
@@ -5236,19 +5236,19 @@ uint64_t RenderingDevice::get_driver_resource(DriverResource p_resource, RID p_r
 			Texture *tex = texture_owner.get_or_null(p_rid);
 			ERR_FAIL_NULL_V(tex, 0);
 
-			driver_id = tex->driver_id;
+			driver_id = tex->driver_id.id;
 		} break;
 		case DRIVER_RESOURCE_SAMPLER: {
 			RDD::SamplerID *sampler_driver_id = sampler_owner.get_or_null(p_rid);
 			ERR_FAIL_NULL_V(sampler_driver_id, 0);
 
-			driver_id = *sampler_driver_id;
+			driver_id = (*sampler_driver_id).id;
 		} break;
 		case DRIVER_RESOURCE_UNIFORM_SET: {
 			UniformSet *uniform_set = uniform_set_owner.get_or_null(p_rid);
 			ERR_FAIL_NULL_V(uniform_set, 0);
 
-			driver_id = uniform_set->driver_id;
+			driver_id = uniform_set->driver_id.id;
 		} break;
 		case DRIVER_RESOURCE_BUFFER: {
 			Buffer *buffer = nullptr;
@@ -5265,19 +5265,19 @@ uint64_t RenderingDevice::get_driver_resource(DriverResource p_resource, RID p_r
 			}
 			ERR_FAIL_NULL_V(buffer, 0);
 
-			driver_id = buffer->driver_id;
+			driver_id = buffer->driver_id.id;
 		} break;
 		case DRIVER_RESOURCE_COMPUTE_PIPELINE: {
 			ComputePipeline *compute_pipeline = compute_pipeline_owner.get_or_null(p_rid);
 			ERR_FAIL_NULL_V(compute_pipeline, 0);
 
-			driver_id = compute_pipeline->driver_id;
+			driver_id = compute_pipeline->driver_id.id;
 		} break;
 		case DRIVER_RESOURCE_RENDER_PIPELINE: {
 			RenderPipeline *render_pipeline = render_pipeline_owner.get_or_null(p_rid);
 			ERR_FAIL_NULL_V(render_pipeline, 0);
 
-			driver_id = render_pipeline->driver_id;
+			driver_id = render_pipeline->driver_id.id;
 		} break;
 		default: {
 			ERR_FAIL_V(0);
diff --git a/servers/rendering/rendering_device_driver.h b/servers/rendering/rendering_device_driver.h
index ef100db97c..e9464ba321 100644
--- a/servers/rendering/rendering_device_driver.h
+++ b/servers/rendering/rendering_device_driver.h
@@ -129,7 +129,6 @@ public:
 #define DEFINE_ID(m_name)                                                                             \
 	struct m_name##ID : public ID {                                                                   \
 		_ALWAYS_INLINE_ explicit operator bool() const { return id != 0; }                            \
-		_ALWAYS_INLINE_ operator uint64_t() const { return id; }                                      \
 		_ALWAYS_INLINE_ m_name##ID &operator=(m_name##ID p_other) {                                   \
 			id = p_other.id;                                                                          \
 			return *this;                                                                             \
diff --git a/servers/rendering/rendering_device_graph.cpp b/servers/rendering/rendering_device_graph.cpp
index 4b85d1c2bf..adac7ee3eb 100644
--- a/servers/rendering/rendering_device_graph.cpp
+++ b/servers/rendering/rendering_device_graph.cpp
@@ -261,7 +261,7 @@ void RenderingDeviceGraph::_add_command_to_graph(ResourceTracker **p_resource_tr
 			}
 
 			if (resource_tracker->parent->usage == RESOURCE_USAGE_NONE) {
-				if (resource_tracker->parent->texture_driver_id != 0) {
+				if (resource_tracker->parent->texture_driver_id.id != 0) {
 					// If the resource is a texture, we transition it entirely to the layout determined by the first slice that uses it.
 					_add_texture_barrier_to_command(resource_tracker->parent->texture_driver_id, RDD::BarrierAccessBits(0), new_usage_access, RDG::RESOURCE_USAGE_NONE, new_resource_usage, resource_tracker->parent->texture_subresources, command_normalization_barriers, r_command->normalization_barrier_index, r_command->normalization_barrier_count);
 				}
@@ -324,7 +324,7 @@ void RenderingDeviceGraph::_add_command_to_graph(ResourceTracker **p_resource_tr
 								ERR_FAIL_MSG("Texture slices that overlap can't be used in the same command.");
 							} else {
 								// Delete the slice from the dirty list and revert it to the usage of the parent.
-								if (current_tracker->texture_driver_id != 0) {
+								if (current_tracker->texture_driver_id.id != 0) {
 									_add_texture_barrier_to_command(current_tracker->texture_driver_id, current_tracker->usage_access, new_usage_access, current_tracker->usage, resource_tracker->parent->usage, current_tracker->texture_subresources, command_normalization_barriers, r_command->normalization_barrier_index, r_command->normalization_barrier_count);
 
 									// Merge the area of the slice with the current tracking area of the command and indicate it's a write usage as well.
@@ -383,7 +383,7 @@ void RenderingDeviceGraph::_add_command_to_graph(ResourceTracker **p_resource_tr
 			while (current_tracker != nullptr) {
 				current_tracker->reset_if_outdated(tracking_frame);
 
-				if (current_tracker->texture_driver_id != 0) {
+				if (current_tracker->texture_driver_id.id != 0) {
 					// Transition all slices to the layout of the parent resource.
 					_add_texture_barrier_to_command(current_tracker->texture_driver_id, current_tracker->usage_access, new_usage_access, current_tracker->usage, resource_tracker->usage, current_tracker->texture_subresources, command_normalization_barriers, r_command->normalization_barrier_index, r_command->normalization_barrier_count);
 				}
```

Let me know if that looks fine, I can update this PR with it (or jsjtxietian can).

Worth noting though, the kind of error you get (at least with GCC) when using `device_id` instead of `device_id.id` is really verbose, and it's quite confusing to assign variables named `device_id` with others that are `blabla_device_id.id`.
This is the error from a single missing `.id`:

```
g++ -o servers/rendering/scu/scu_servers_rendering.gen.linuxbsd.editor.dev.x86_64.o -c -std=gnu++17 -fno-exceptions -Wctor-dtor-privacy -Wnon-virtual-dtor -Wplacement-new=1 -pipe -gdwarf-4 -g3 -O0 -Wall -Wextra -Wwrite-strings -Wno-unused-parameter -Wshadow -Wno-misleading-indentation -Wno-return-type -Walloc-zero -Wduplicated-branches -Wduplicated-cond -Wstringop-overflow=4 -Wattribute-alias=2 -Wlogical-op -Werror -isystem thirdparty/glad -DTOOLS_ENABLED -DDEBUG_ENABLED -DDEV_ENABLED -DNO_EDITOR_SPLASH -DSOWRAP_ENABLED -DTOUCH_ENABLED -DFONTCONFIG_ENABLED -DALSA_ENABLED -DALSAMIDI_ENABLED -DPULSEAUDIO_ENABLED -D_REENTRANT -DDBUS_ENABLED -DSPEECHD_ENABLED -DXKB_ENABLED -DJOYDEV_ENABLED -DUDEV_ENABLED -DLINUXBSD_ENABLED -DUNIX_ENABLED -D_FILE_OFFSET_BITS=64 -DX11_ENABLED -DLIBDECOR_ENABLED -DWAYLAND_ENABLED -DVULKAN_ENABLED -DRD_ENABLED -DGLES3_ENABLED -DCRASH_HANDLER_ENABLED -DMINIZIP_ENABLED -DBROTLI_ENABLED -DTHREADS_ENABLED -DCLIPPER2_ENABLED -DZSTD_STATIC_LINKING_ONLY -DUSE_VOLK -DVK_USE_PLATFORM_XLIB_KHR -DVK_USE_PLATFORM_WAYLAND_KHR -DGLAD_ENABLED -DEGL_ENABLED -DGLAD_GLX_NO_X11 -DHAVE_MNTENT -Ithirdparty/freetype/include -Ithirdparty/libpng -Ithirdparty/volk -Ithirdparty/vulkan -Ithirdparty/vulkan/include -Ithirdparty/zstd -Ithirdparty/zlib -Ithirdparty/clipper2/include -Ithirdparty/brotli/include -Iplatform/linuxbsd -Ithirdparty/linuxbsd_headers/wayland -Ithirdparty/linuxbsd_headers -Iplatform/linuxbsd -I. servers/rendering/scu/scu_servers_rendering.gen.cpp
In file included from servers/rendering/scu/scu_servers_rendering.gen.cpp:14:
./servers/rendering/rendering_device_graph.cpp: In member function 'void RenderingDeviceGraph::_add_command_to_graph(ResourceTracker**, ResourceUsage*, uint32_t, int32_t, RecordedCommand*)':
./servers/rendering/rendering_device_graph.cpp:386:72: error: no match for 'operator!=' (operand types are 'RenderingDeviceDriver::TextureID' and 'int')
  386 |                                 if (current_tracker->texture_driver_id != 0) {
      |                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^~ ~
      |                                                      |                    |
      |                                                      |                    int
      |                                                      RenderingDeviceDriver::TextureID
./servers/rendering/rendering_device_graph.cpp:386:72: note: candidate: 'operator!=(int, int)' (built-in)
  386 |                                 if (current_tracker->texture_driver_id != 0) {
      |                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
./servers/rendering/rendering_device_graph.cpp:386:72: note:   no known conversion for argument 1 from 'RenderingDeviceDriver::TextureID' to 'int'
In file included from ./servers/rendering/rendering_device.h:43,
                 from ./servers/rendering_server.h:42,
                 from ./servers/rendering/environment/renderer_fog.h:34,
                 from ./servers/rendering/renderer_compositor.h:34,
                 from ./servers/rendering/renderer_canvas_cull.h:35,
                 from ./servers/rendering/renderer_canvas_cull.cpp:31,
                 from servers/rendering/scu/scu_servers_rendering.gen.cpp:1:
./servers/rendering/rendering_device_driver.h:138:38: note: candidate: 'bool RenderingDeviceDriver::TextureID::operator!=(const RenderingDeviceDriver::TextureID&) const'
  138 |                 _ALWAYS_INLINE_ bool operator!=(const m_name##ID &p_other) const { return id != p_other.id; } \
      |                                      ^~~~~~~~
./servers/rendering/rendering_device_driver.h:149:9: note: in expansion of macro 'DEFINE_ID'
  149 |         DEFINE_ID(Texture);
      |         ^~~~~~~~~
./servers/rendering/rendering_device_driver.h:138:67: note:   no known conversion for argument 1 from 'int' to 'const RenderingDeviceDriver::TextureID&'
  138 |                 _ALWAYS_INLINE_ bool operator!=(const m_name##ID &p_other) const { return id != p_other.id; } \
      |                                                 ~~~~~~~~~~~~~~~~~~^~~~~~~
./servers/rendering/rendering_device_driver.h:149:9: note: in expansion of macro 'DEFINE_ID'
  149 |         DEFINE_ID(Texture);
      |         ^~~~~~~~~
In file included from ./core/templates/hash_map.h:38,
                 from ./core/object/object.h:39,
                 from ./core/variant/binder_common.h:35,
                 from ./core/object/method_bind.h:34,
                 from ./core/object/class_db.h:34,
                 from ./core/object/ref_counted.h:34,
                 from ./core/io/resource_uid.h:34,
                 from ./core/io/resource.h:34,
                 from ./core/io/image.h:34,
                 from ./servers/rendering_server.h:34:
./core/templates/pair.h:58:6: note: candidate: 'template<class F, class S> bool operator!=(const Pair<F, S>&, const Pair<F, S>&)'
   58 | bool operator!=(const Pair<F, S> &pair, const Pair<F, S> &other) {
      |      ^~~~~~~~
./core/templates/pair.h:58:6: note:   template argument deduction/substitution failed:
./servers/rendering/rendering_device_graph.cpp:386:75: note:   'RenderingDeviceDriver::TextureID' is not derived from 'const Pair<F, S>'
  386 |                                 if (current_tracker->texture_driver_id != 0) {
      |                                                                           ^
./core/templates/pair.h:103:6: note: candidate: 'template<class K, class V> bool operator!=(const KeyValue<K, V>&, const KeyValue<K, V>&)'
  103 | bool operator!=(const KeyValue<K, V> &pair, const KeyValue<K, V> &other) {
      |      ^~~~~~~~
./core/templates/pair.h:103:6: note:   template argument deduction/substitution failed:
./servers/rendering/rendering_device_graph.cpp:386:75: note:   'RenderingDeviceDriver::TextureID' is not derived from 'const KeyValue<K, V>'
  386 |                                 if (current_tracker->texture_driver_id != 0) {
      |                                                                           ^
In file included from ./core/templates/paged_allocator.h:37,
                 from ./servers/rendering/renderer_canvas_cull.h:34:
./core/string/ustring.h:481:6: note: candidate: 'bool operator!=(const char*, const String&)'
  481 | bool operator!=(const char *p_chr, const String &p_str);
      |      ^~~~~~~~
./core/string/ustring.h:481:29: note:   no known conversion for argument 1 from 'RenderingDeviceDriver::TextureID' to 'const char*'
  481 | bool operator!=(const char *p_chr, const String &p_str);
      |                 ~~~~~~~~~~~~^~~~~
./core/string/ustring.h:482:6: note: candidate: 'bool operator!=(const wchar_t*, const String&)'
  482 | bool operator!=(const wchar_t *p_chr, const String &p_str);
      |      ^~~~~~~~
./core/string/ustring.h:482:32: note:   no known conversion for argument 1 from 'RenderingDeviceDriver::TextureID' to 'const wchar_t*'
  482 | bool operator!=(const wchar_t *p_chr, const String &p_str);
      |                 ~~~~~~~~~~~~~~~^~~~~
In file included from ./core/string/node_path.h:34,
                 from ./core/variant/variant.h:55,
                 from ./core/object/message_queue.h:38,
                 from ./core/object/object.h:35:
./core/string/string_name.h:198:6: note: candidate: 'bool operator!=(const String&, const StringName&)'
  198 | bool operator!=(const String &p_name, const StringName &p_string_name);
      |      ^~~~~~~~
./core/string/string_name.h:198:31: note:   no known conversion for argument 1 from 'RenderingDeviceDriver::TextureID' to 'const String&'
  198 | bool operator!=(const String &p_name, const StringName &p_string_name);
      |                 ~~~~~~~~~~~~~~^~~~~~
./core/string/string_name.h:200:6: note: candidate: 'bool operator!=(const char*, const StringName&)'
  200 | bool operator!=(const char *p_name, const StringName &p_string_name);
      |      ^~~~~~~~
./core/string/string_name.h:200:29: note:   no known conversion for argument 1 from 'RenderingDeviceDriver::TextureID' to 'const char*'
  200 | bool operator!=(const char *p_name, const StringName &p_string_name);
      |                 ~~~~~~~~~~~~^~~~~~
In file included from ./core/math/delaunay_3d.h:43,
                 from ./core/math/geometry_3d.h:34,
                 from ./servers/rendering_server.h:35:
./thirdparty/misc/r128.h:530:20: note: candidate: 'bool operator!=(const R128&, const R128&)'
  530 | static inline bool operator!=(const R128 &lhs, const R128 &rhs)
      |                    ^~~~~~~~
./thirdparty/misc/r128.h:530:43: note:   no known conversion for argument 1 from 'RenderingDeviceDriver::TextureID' to 'const R128&'
  530 | static inline bool operator!=(const R128 &lhs, const R128 &rhs)
      |                               ~~~~~~~~~~~~^~~
scons: *** [servers/rendering/scu/scu_servers_rendering.gen.linuxbsd.editor.dev.x86_64.o] Error 1
```

It tries a lot of obscure conversions, maybe there are ways to prevent those to get a cleaner error message with less `note`s.

Reply:
> Here's the diff needed to make it compile (at least Linux editor dev build) without that operator:

Thanks for the help, now is China's holiday so I did not reply yesterday.

> The way this was designed, `.id` should be used explicitly.

I was thinking can we remove the operator bool() and operator uint64_t() since we want explict, and apply the patch. Inexplict always causes trouble lol.

Reply:
For the time being, I'd go with the patch as well as the explicit bool operator. The operator is important for fixing silent issues, so I'd say that's superior despite the issues with only it (without the other conversion operator) are quite verbose and hard-to-read. We may be able to come up with something that has the best of the three world afterwards.

Reply:
> I'd go with the patch as well as the explicit bool operator.

Done.